### PR TITLE
invariant shrink #6683: check if test failed instead revert

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation

see https://github.com/foundry-rs/foundry/issues/6683#issuecomment-1966794329
Atm shrinking doesn't apply for invariants using asserts but only for invariants that reverts (require)

## Solution

- when deciding if a sequence can fail test check if test failed (instead if reverted only)
- added test with assert and require to ensure they shrink at same len

### CODE REVIEW

File: crates/evm/evm/src/executors/invariant/error.rs [hunk 1]
```diff
@@ -214,12 +214,18 @@ impl InvariantFuzzError {
...
+            // Checks the invariant. If we revert or fail before the last call, all the better.
             if let Some(func) = &self.func {
-                let error_call_result = executor
+                let call_result = executor
                     .call_raw(CALLER, self.addr, func.clone(), U256::ZERO)
                     .expect("bad call to evm");
-                if error_call_result.reverted {
+                let is_success = executor.is_raw_call_success(
+                    self.addr,
+                    call_result.state_changeset.clone().unwrap(),
```
Comment (mattsse) [inline crates/evm/evm/src/executors/invariant/error.rs:224]: do we need the clone here?

**Review by mattsse — CHANGES_REQUESTED** [review]
I believe this makes sense,

one q re clone
+ lint issue

File: crates/evm/evm/src/executors/invariant/error.rs (hunk 1 above)
Comment (grandizzy) [reply crates/evm/evm/src/executors/invariant/error.rs:224]: the next param is `&call_result` which borrows the value so  if not cloning fails with
```
225 |                     &call_result,
    |                     ^^^^^^^^^^^^ value borrowed here after partial move
```

Comment (mattsse) [reply crates/evm/evm/src/executors/invariant/error.rs:224]: hmm, the `pub fn is_raw_call_success(` fn is not very good, I see why we need the state as a separate arg

but this means for this call, we can use `mut call_result.state_changeset.take().unwrap()`

Comment (grandizzy) [reply crates/evm/evm/src/executors/invariant/error.rs]: thank you, this makes sense, I made following change in https://github.com/foundry-rs/foundry/pull/7257/commits/9d2b073c4a3d9afb4fd53464b76b5bd28a179177 + made test more robust as they can change in less steps

**Review by mattsse — APPROVED** [review]

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**